x-pack/metricbeat/module/aws/cloudwatch/cloudwatch.go

```diff
@@ -0,0 +1,326 @@
+		} else {
+			namespacesTotal = append(namespacesTotal, namespace)
+		}
+	}
+	return listMetricsTotal, namespacesTotal
+}
+
+func constructMetricQueries(listMetricsOutput []cloudwatch.Metric, period int64) []cloudwatch.MetricDataQuery {
+	metricDataQueries := []cloudwatch.MetricDataQuery{}
+	for i, listMetric := range listMetricsOutput {
+		metricDataQuery := createMetricDataQuery(listMetric, i, period)
+		metricDataQueries = append(metricDataQueries, metricDataQuery)
+	}
+	return metricDataQueries
+}
+
+func constructLabel(metric cloudwatch.Metric) string {
+	metricDims := metric.Dimensions
+	metricName := *metric.MetricName
+	label := metricName + " " + *metric.Namespace
+	dimNames := ""
+	dimValues := ""
+	for i, dim := range metricDims {
+		dimNames += *dim.Name
+		dimValues += *dim.Value
+		if i != len(metricDims)-1 {
+			dimNames += ","
+			dimValues += ","
+		}
+	}
+
+	if dimNames != "" && dimValues != "" {
+		label += " " + dimNames
+		label += " " + dimValues
+	}
+	return label
+}
+
+func createMetricDataQuery(metric cloudwatch.Metric, index int, period int64) (metricDataQuery cloudwatch.MetricDataQuery) {
+	statistic := "Average"
+	id := "cw" + strconv.Itoa(index)
+	label := constructLabel(metric)
+
+	metricDataQuery = cloudwatch.MetricDataQuery{
+		Id: &id,
+		MetricStat: &cloudwatch.MetricStat{
+			Period: &period,
+			Stat:   &statistic,
+			Metric: &metric,
+		},
+		Label: &label,
+	}
+	return
+}
+
+func getIdentifiers(listMetricsOutputs []cloudwatch.Metric) map[string][]string {
+	if len(listMetricsOutputs) == 0 {
+		return nil
+	}
+
+	identifiers := map[string][]string{}
+	for _, listMetrics := range listMetricsOutputs {
+		identifierName := ""
+		identifierValue := ""
+		if len(listMetrics.Dimensions) == 0 {
+			continue
+		}
+
+		for i, dim := range listMetrics.Dimensions {
+			identifierName += *dim.Name
+			identifierValue += *dim.Value
+			if i != len(listMetrics.Dimensions)-1 {
+				identifierName += ","
+				identifierValue += ","
+			}
+		}
+
+		if identifiers[identifierName] != nil {
+			if !aws.StringInSlice(identifierValue, identifiers[identifierName]) {
+				identifiers[identifierName] = append(identifiers[identifierName], identifierValue)
+			}
+		} else {
+			identifiers[identifierName] = []string{identifierValue}
+		}
+	}
+
+	return identifiers
+}
+
+func initEvent(regionName string) mb.Event {
+	event := mb.Event{}
+	event.Service = metricsetName
+	event.RootFields = common.MapStr{}
+	event.MetricSetFields = common.MapStr{}
+	event.RootFields.Put("service.name", metricsetName)
+	if regionName != "" {
+		event.RootFields.Put("cloud.region", regionName)
+	}
+	return event
+}
+
+func insertMetricSetFields(event mb.Event, metricValue float64, labels []string) mb.Event {
+	event.MetricSetFields.Put("namespace", labels[1])
+	event.MetricSetFields.Put(labels[0], metricValue)
+	if len(labels) == 2 {
+		return event
+	}
+
+	dimNames := strings.Split(labels[2], ",")
+	dimValues := strings.Split(labels[3], ",")
```

Comment:
The position of these labels seems meaningful, could we use a better data structure as a list of structs to pass them as arguments? Or if a slice is the best option, use some constants for the indexes.

Reply:
Yeah, unfortunately we have to use a string here to be the label. So I will use constants for the index (till I think of something better 😄). Thanks!